### PR TITLE
disable in place deployment for function app

### PR DESCRIPTION
**Pull request by watashiSHUN**
add a simple patch to the existing function`CopyDirectoryRecursive` to exclude directories/subdirectories with name "node_modules". This will cover functionApp's use case where there's a `package.json` under `wwwroot` or under each `wwwroot/functionFolder`

note to self:
1. add `SCM_USE_FUNCPACK` to kudu doc as a deployment flag
2. add functional tests for git clone functionApp

### CODE REVIEW

File: Kudu.Services/GitServer/InfoRefsController.cs [hunk 1]
```diff
@@ -185,11 +187,29 @@ public void InitialCommitIfNecessary()
...
 
-                // do initial commit
-                repository = _repositoryFactory.EnsureRepository(RepositoryType.Git);
+                // if it is a functionApp
+                // git clone will copy repo from wwwroot/ to repository/
+                // then treate it like its not in place deployment
+                if (!string.IsNullOrEmpty(System.Environment.GetEnvironmentVariable(Constants.FunctionRunTimeVersion)))
+                {
+                    // since git clone doesn't happen very often, create this dynamically instead of static
+                    var ignoreDir = new HashSet<string>();
```
Comment (ahmelsayed) [inline Kudu.Services/GitServer/InfoRefsController.cs]: maybe use a `HashSet<string>(StringComparer.OrdinalIgnoreCase)` is OS is Windows?

File: Kudu.Core/Infrastructure/FileSystemHelpers.cs [hunk 2]
```diff
@@ -201,7 +201,8 @@ public static void MoveFile(string sourceFileName, string destFileName)
         }
 
         // From MSDN: http://msdn.microsoft.com/en-us/library/bb762914.aspx
-        public static void CopyDirectoryRecursive(string sourceDirPath, string destinationDirPath, bool overwrite = true)
+        // <param name="ignoreDir">string of directory/subdirectory names to be excluded from the copy</param>
+        public static void CopyDirectoryRecursive(string sourceDirPath, string destinationDirPath, bool overwrite = true, HashSet<string> ignoreDir = null)
```
Comment (ahmelsayed) [inline Kudu.Core/Infrastructure/FileSystemHelpers.cs]: nit: ignoreDir**s**?

**Comment by watashiSHUN** [comment]
added unit test

**Comment by jvano** [comment]
Hi

If the problem persists and is related to running it on Azure App Service, please open a support incident in Azure:
https://learn.microsoft.com/en-us/azure/azure-portal/supportability/how-to-create-azure-support-request

This way we can better track and assist you on this case

Thanks,

Joaquin Vano
Azure App Service

**Pull request closed without merging**